### PR TITLE
NPSP Cascade Deletion Redesign

### LINKED ISSUE #2377: Recurring Donation is not deleted if Account/Contact are deleted
We are seeing users who are getting NPSP Error emails with the error "Attempt to de-reference a null object"

This appears to consistently be because the Organization and Contact are blank on a Recurring Donation record.

I don't believe this is bad data entry. When you insert a Recurring Donation without an Organization or Contact, you get an error. 

This is a consistent way to reproduce the error and may be why our users are hitting it - when an Account (or Contact) connected to the Recurring Donation is deleted, the Recurring Donation is left behind! The nightly batch is attempting to roll up that now orphaned Recurring Donation record and getting the exception.

Steps to repeat:

1. Create Organization Account (or Contact)
2. Create Recurring Donation, open, $1, Yearly Installment Period (so there's only 1 opp created)
3. Delete Organization Account (or Contact)
4. Run Recurring Donation Rollup - get 'Attempt to de-reference a null object."

Recurring Donation is not deleted as part of the cascade delete of Account/Contact (even though the Opportunities are deleted). 

Delete behavior should remove orphaned Recurring Donation, and/or those records should not be calculated in the batch update.

### LINKED ISSUE #2443: Delete reciprocal Relationship records when Contact is deleted
thread: https://powerofus.force.com/0D580000039cpMZ

Issue replicated in test org.

### LINKED ISSUE #2449: When Contacts are merged with Recurring Donations, the losing Contact's Recurring Donation(s) are deleted
Thread: https://powerofus.force.com/0D58000003AUWPg

Opportunities are retained, it's only the Recurring Donation that's deleted. It's recoverable from recycle bin.

### PR DESCRIPTION
**Pull request by njjc**
# Critical Changes

# Changes
- Deleting a Contact now cascade delete's that contact's Opportunities via the Primary Contact field. If the Contact has any Closed Won or Organization attributed opportunities, an error prevents deletion.
- Deleting a Recurring Donation with associated Opportunities no longer causes a validation error, but instead cascade deletes those Opportunities. Cascade deleted Opportunities show up in the recycle bin and can be undeleted separately from the Recurring Donation.

# Issues Closed
- Fixes #2377 
- Fixes #2443 
- Fixes #2449

### CODE REVIEW

**Comment by davidhabib** [comment]
@njjc didn't we have a chatter thread going on what behavior we wanted?  I don't recall seeing a final consensus on this.  did you have a conversation with others that I missed?  thanks.

**Comment by njjc** [comment]
I didn't, as the other relevant parties were out and I wanted to make progress. We can reevaluate my conclusions together and with the rest of the team.

File: src/classes/ACCT_IndividualAccounts_TDTM.cls [hunk 1]
```diff
@@ -665,8 +665,10 @@ public class ACCT_IndividualAccounts_TDTM extends TDTM_Runnable {
         }
         // delete the accounts
         if (accountsForDeletion.size() > 0) {
-            dmlWrapper.objectsToDelete.addAll(accountsForDeletion);           
-            //delete accountsForDeletion;
+            /*This is a bad strategy as it leaves the deleted contact unrecoverable from the recycle bin.
+            **Leaving an orphaned account is less of an issue than unrecoverable data, and that is solved
+            **by the contact delete button override. */
+            //dmlWrapper.objectsToDelete.addAll(accountsForDeletion);
```
Comment (njjc) [inline src/classes/ACCT_IndividualAccounts_TDTM.cls]: Exception on firing contact is needed here to encourage enabling the delete override.

**Comment by njjc** [comment]
@davidhabib this is ready for review. There are a couple of test failures with the data importer that's expecting the old behavior of auto deleting orphaned accounts; I could use your help in figuring out if we need to update the code or just the tests to fix those errors. Thanks!

**Comment by njjc** [comment]
@jlantz how can I rerun the build for [this branch](https://mrbelvedereci.herokuapp.com/repo/SalesforceFoundation/Cumulus/branch/feature/2377-rd-exception) in the new front end?

**Pull request closed without merging**